
<img src="res/viu_logo.png" width="200">

## 01MIAR - Actividad WhitePaper
### Nombre: Mayra Pullupaxi


#### Objetivo
- Leer y analizar críticamente dos artículos científicos.

### Artículo 01


### The NumPy array: a structure for efficient numerical computation

#### Actividad 01.01 - Ampliar dicha explicación, aportando posibles restricciones o limitaciones a dicho sistema y ejemplos propios de los casos de uso.

El Broadcasting permite operar con arreglos de diferentes dimensiones, de esta manera realiza cálculos aritméticos en un lenguaje de alto nivel y mejora el rendimiento mediante tres técnicas: vectorizando los cálculos, evitando copiar en memoria y minimizando las operaciones. A continuación un ejemplo:

In [1]:
import numpy as np

In [9]:
a = np.arange(8).reshape(4, 2)
print(a)
b = np.arange(2).reshape(1, 2)
print(b)
print(a * b)

[[0 1]
 [2 3]
 [4 5]
 [6 7]]
[[0 1]]
[[0 1]
 [0 3]
 [0 5]
 [0 7]]


La restricción es que para realizar los cálculos aritméticos es necesario que las matrices sean de igual dimensión de elemento en elemento sino será un error, entonces la limitación es que las matrices deben ser iguales en su forma y son compatibles cuando: son iguales o uno es 1.

In [10]:
#Matriz de dimension 4 x 3
a = np.arange(12).reshape(4, 3)
print(a)

#Matriz de dimension 1 x 2
b = np.arange(2).reshape(1, 2)
print(b)

print(a * b)

[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]
[[0 1]]


ValueError: operands could not be broadcast together with shapes (4,3) (1,2) 

En este ejemplo se muestra la limitación, al parecer las operaciones funcionarían, pero cuando se combinan y las dimensiones en cuanto a la forma no es misma existe error.

In [15]:
a = np.random.randn(9,1)
b = np.random.randn(1,9)
c = np.random.randn(9,12)

print(np.dot(a,b))

d = np.dot(a,b) * c

[[-0.60517948  0.45853794  1.13486184  0.99947507 -0.28382997  0.34869326
   0.85624111 -0.38128319 -0.1620085 ]
 [-0.54544185  0.41327538  1.02283894  0.90081628 -0.25581294  0.31427354
   0.77172103 -0.3436465  -0.14601654]
 [-0.9626993   0.72942683  1.8053003   1.58993155 -0.45150724  0.55468959
   1.36207975 -0.60653256 -0.2577177 ]
 [ 0.82399914 -0.62433522 -1.54520305 -1.36086338  0.38645668 -0.47477311
  -1.16583916  0.51914685  0.22058722]
 [-0.12281115  0.09305268  0.23030142  0.20282691 -0.05759859  0.07076152
   0.17375995 -0.07737511 -0.03287694]
 [ 0.59108808 -0.44786104 -1.10843698 -0.97620264  0.27722109 -0.34057406
  -0.83630383  0.37240514  0.15823618]
 [-0.27990974  0.21208458  0.5249003   0.46228073 -0.13127804  0.16127883
   0.39603165 -0.17635244 -0.07493274]
 [-0.41677901  0.31578894  0.78156417  0.68832511 -0.1954699   0.24014038
   0.5896818  -0.26258463 -0.11157308]
 [ 0.55135016 -0.41775204 -1.03391849 -0.91057406  0.25858395 -0.3176778
  -0.78008043  0.347368

ValueError: operands could not be broadcast together with shapes (9,9) (9,12) 

#### Actividad 01.02 - Verificar la eficacia y mejora posible de rendimiento del uso de dicha técnica sobre ndarrays de tamaños grandes.

En la ficha trata sobre memmap que son objetos en forma de matriz, además crea un mapa de memoria para una matriz almacenada en un archivo grande del disco.
Esto permite que sin leer todo el archivo en la memoria se acceda a partes del archivo en el disco.

Siempre se debe primero vaciar la instancia de memmap para escribir los cambios en el archivo, ya que todavía no existe un proceso para que lo haga. Y así asegurarse que el recurso esté realmente cerrado, ya que puede compartirse entre diferentes instancias de memmap.

In [29]:
import numpy as np
from time import time 

from tempfile import mkdtemp
import os.path as path
ruta = path.join(mkdtemp(), 'data.dat')

In [30]:
data_array = np.memmap(ruta, dtype = np.float64, mode = 'w+', shape=(100, 100, 100))

t = time() 
for i in range(5): 
    view = np.array(data_array[::16, ::16, ::16]) 
t = ((time() - t)/5) * 1000 
print("Time (ms): %i" % t )

Time (ms): 1


### Artículo 02

#### Actividad 02.01 - Desarrollar una opinión razonada del estado actual de las herramientas de análisis de datos estadísticos en contraposición a como se muestran en el artículo, R vs Python vs SQL vs Others...

Si bien Python ha dado grandes pasos últimamente en el aprendizaje automático, se ha estado quedando con el análisis estadístico de datos, que en comparación con R que tiene amplitud en bibliotecas de código abierto bien examinadas y están disponibles para los usuarios y también en comparación con lenguajes especializados como Matlab, falta un nivel de madurez y estabilidad de las bibliotecas.

Una de las limitaciones que tiene Pandas en Python es que no permite trabajar con tipos de datos mixtos, pero si se puede usar con gran cantidad de datos que será necesario validar sino se compromete el rendimiento. Python al ser un lenguaje de alto nivel es mucho más rápido que R puesto que compila en tiempo de ejecución, mientras que R es programación compilada.

Para saber si usar Python o R es necesario conocer en donde va a ser aplicado el análisis estadístico ya que si se requiere un análisis exhaustivo de datos con amplios conocimientos matemáticos y estadísticos el lenguaje que a utilizar debería ser R, y a propósito es muy usado por los matemáticos. En cambio cuando se tiene conocimientos en programación y en informática será más fácil utilizar Python ya que su propósito es más general.

Python es un lenguaje de programación de uso general mientras que SQL es un lenguaje de propósito especial usado para acceder a los datos y cada uno tiene su fin. Dentro de unos años más se verá los avances de Python frente a R y se podrá realizar una nueva comprativa. 

